### ISSUE TITLE
wav2vec_ctc valid_wer keeps 100 ?

### ISSUE THREAD

**Issue by apoca909** [pytorch/fairseq#3340]
I finetune wav2vec_ctc model on chinese mandarian audio data, while the valid_wer always 100

train log
[hydra_train.log](https://github.com/pytorch/fairseq/files/6128370/hydra_train.log)

debug in inference code
#debug in W2lKenLMDecoder.deocde

p nbest_results[0].words
[-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1]

(Pdb) p emissions.shape
torch.Size([1, 149, 7210])

and the emissions' argsmax values are 4s

yaml:

common:
  fp16: true
  log_format: json
  log_interval: 200

checkpoint:
  no_epoch_checkpoints: true
  best_checkpoint_metric: wer
  maximize_best_checkpoint_metric: false
 
task:
  _name: audio_pretraining
  data: ???
  normalize: false
  labels: ltr

dataset:
  num_workers: 6
  max_tokens: 3200000
  skip_invalid_size_inputs_valid_test: true
  valid_subset: valid

distributed_training:
  ddp_backend: no_c10d
  distributed_world_size: 8

criterion:
  _name: ctc
  zero_infinity: true

optimization:
  max_update: 320000
  lr: [0.00001]
  sentence_avg: true

optimizer:
  _name: adam
  adam_betas: (0.9,0.98)
  adam_eps: 1e-08

lr_scheduler:
  _name: tri_stage
  phase_ratio: [0.1, 0.4, 0.5]
  final_lr_scale: 0.05

model:
  _name: wav2vec_ctc
  w2v_path: ???
  apply_mask: true
  mask_prob: 0.4
  mask_channel_prob: 0.1
  mask_channel_length: 64
  layerdrop: 0.1
  activation_dropout: 0.1
  feature_grad_mult: 0.0
  freeze_finetune_updates: 0

any ideas ?

**Comment by apoca909** [comment]
maybe the dict size (7210)  too large?

**Comment by MGLiXu** [comment]
> maybe the dict size 7210 too large?
> i'm trying convert mandarin char to pinyin , and the dict size down to 39, i will give the finetune result later.

hi,
I encounter the same problem. The reason perhaps  is related to the hyper parameter "freeze_finetune_updates". For me, after 3000 epochs(base_1h.yaml config), the WER begins to decrease. 
But I have another question. I use this model to process my audio, the result is as follows:
然 ， 然 后 您 您 这 边 ， 就 是 需 要 ， 发 发 给 我 ， 这 个 ， 情 点 报 告 ， 我 帮 ， 你 去 看 ， 病 ， ， ， ， ， ， 啊 ， 到 ， 给 ， 这 些 ， 一 些 ， 健 康 问 题 ， 需 要 告 告 知 ， ， ， ， ， 然 有 有 没 有 影 影 响 。 。 。
There are too many repeated character. I don't know if this is related to the dataset format. I treat every character as a word. Should I split the words first and then train this model. Do you have any ideas on that?  What is your result of using pinyin? Can you share it?

**Comment by apoca909** [comment]
> > maybe the dict size 7210 too large?
> > i'm trying convert mandarin char to pinyin , and the dict size down to 39, i will give the finetune result later.
> 
> hi,
> I encounter the same problem. The reason perhaps is related to the hyper parameter "freeze_finetune_updates". For me, after 3000 epochs(base_1h.yaml config), the WER begins to decrease.
> But I have another question. I use this model to process my audio, the result is as follows:
> 然 ， 然 后 您 您 这 边 ， 就 是 需 要 ， 发 发 给 我 ， 这 个 ， 情 点 报 告 ， 我 帮 ， 你 去 看 ， 病 ， ， ， ， ， ， 啊 ， 到 ， 给 ， 这 些 ， 一 些 ， 健 康 问 题 ， 需 要 告 告 知 ， ， ， ， ， 然 有 有 没 有 影 影 响 。 。 。
> There are too many repeated character. I don't know if this is related to the dataset format. I treat every character as a word. Should I split the words first and then train this model. Do you have any ideas on that? What is your result of using pinyin? Can you share it?


the param freeze_finetune_updates have set to 0. maybe the finetuning model is underfitting .
i have about 4k hours data for finetuning, and for lack of GPUs, the training is slow, some log as follows:

[2021-03-16 00:47:12,746][valid][INFO] - {"epoch": 5, "valid_loss": "16.003", "valid_ntokens": "2935.52", "valid_nsentences": "94.8571", "valid_nll_loss": "0.517", "valid_uer": "11.647", "valid_wer": "28.432", "valid_raw_wer": "41.176", "valid_wps": "1791.9", "valid_wpb": "2935.5", "valid_bsz": "94.9", "valid_num_updates": "35864", "valid_best_wer": "28.432"}
[2021-03-16 00:47:13,819][fairseq.checkpoint_utils][INFO] - Preparing to save checkpoint for epoch 5 @ 35864 updates
[2021-03-16 00:47:30,257][fairseq.trainer][INFO] - Finished saving checkpoint to checkpoints/checkpoint_best.pt
[2021-03-16 00:48:10,979][fairseq.checkpoint_utils][INFO] - Saved checkpoint checkpoints/checkpoint_best.pt (epoch 5 @ 35864 updates, score 28.432) (writing took 57.159094674978405 seconds)
[2021-03-16 00:48:11,980][fairseq_cli.train][INFO] - end of epoch 5 (average epoch stats below)
[2021-03-16 00:48:11,999][train][INFO] - {"epoch": 5, "train_loss": "26.089", "train_ntokens": "17873.6", "train_nsentences": "578.432", "train_nll_loss": "0.844", "train_wps": "3978.8", "train_ups": "0.22", "train_wpb": "17873.6", "train_bsz": "578.4", "train_num_updates": "35864", "train_lr": "1e-05", "train_gnorm": "18.944", "train_loss_scale": "16", "train_train_wall": "23778", "train_wall": "159605"}
[2021-03-16 00:48:12,118][fairseq.trainer][INFO] - begin training epoch 6
[2021-03-16 01:15:08,734][train_inner][INFO] - {"epoch": 6, "update": 5.019, "loss": "25.133", "ntokens": "17728", "nsentences": "573.87", "nll_loss": "0.814", "wps": "1489.3", "ups": "0.08", "wpb": "17728", "bsz": "573.9", "num_updates": "36000", "lr": "1e-05", "gnorm": "19.43", "loss_scale": "16", "train_wall": "536", "wall": "161222"}
[2021-03-16 01:23:57,240][train_inner][INFO] - {"epoch": 6, "update": 5.047, "loss": "25.047", "ntokens": "17910.3", "nsentences": "581.72", "nll_loss": "0.814", "wps": "6780.8", "ups": "0.38", "wpb": "17910.3", "bsz": "581.7", "num_updates": "36200", "lr": "1e-05", "gnorm": "19", "loss_scale": "16", "train_wall": "444", "wall": "161750"}
[2021-03-16 01:33:10,504][train_inner][INFO] - {"epoch": 6, "update": 5.075, "loss": "25.268", "ntokens": "17862", "nsentences": "577.07", "nll_loss": "0.816", "wps": "6467.3", "ups": "0.36", "wpb": "17862", "bsz": "577.1", "num_updates": "36400", "lr": "1e-05", "gnorm": "19.291", "loss_scale": "32", "train_wall": "388", "wall": "162304"}
[2021-03-16 01:36:09,957][fairseq.trainer][INFO] - NOTE: gradient overflow detected, ignoring gradient, setting loss scale to: 16.0

**Comment by apoca909** [comment]
> > maybe the dict size 7210 too large?
> > i'm trying convert mandarin char to pinyin , and the dict size down to 39, i will give the finetune result later.
> 
> hi,
> I encounter the same problem. The reason perhaps is related to the hyper parameter "freeze_finetune_updates". For me, after 3000 epochs(base_1h.yaml config), the WER begins to decrease.
> But I have another question. I use this model to process my audio, the result is as follows:
> 然 ， 然 后 您 您 这 边 ， 就 是 需 要 ， 发 发 给 我 ， 这 个 ， 情 点 报 告 ， 我 帮 ， 你 去 看 ， 病 ， ， ， ， ， ， 啊 ， 到 ， 给 ， 这 些 ， 一 些 ， 健 康 问 题 ， 需 要 告 告 知 ， ， ， ， ， 然 有 有 没 有 影 影 响 。 。 。
> There are too many repeated character. I don't know if this is related to the dataset format. I treat every character as a word. Should I split the words first and then train this model. Do you have any ideas on that? What is your result of using pinyin? Can you share it?

改拼音的方法跟音素的方法本质上是相同的，对于带有口音、方言的音频可能效果并不好，好处是可以降低字典规模，我最近用了BPE的字典编码方案，字典规模缩小了很多，单看结果似乎还行，中间部分结果如下(不过开发集的wer 刷到17之后就基本上不去了，ngram LM)：
{"epoch": 13, "update": 12.022, "loss": "26.584", "ntokens": "17028.9", "nsentences": "724.595", "nll_loss": "1.131", "wps": "1519", "ups": "0.09", "wpb": "17028.9", "bsz": "724.6", "num_updates": "68400", "lr": "1e-06", "gnorm": "76.588", "loss_scale": "0.25", "train_wall": "573", "wall": "0"}

你觉得我妈妈美不美胖不胖 (None-2463)
下一首下一首下一首 (None-316)
我给你唰两那个你你你你没有的音乐我想天怎么办呢 (None-3349)
可是我不我喜不喜欢喝你奶我喜欢喝舴奢 (None-2502)
每周一到周五七点二十叫我朋友起来上学 (None-2150)
七十五分七十五分之四十四百九十乘以二十一分之二十等于多少 (None-3748)
放大王叫我来巡山小度小度小度小度 (None-2310)
苹果英语怎么说 (None-1234)

有些识别中带有错字，还在排查，可能是标注数据本身有错误，也可能是语言模型数据不足。。
还有想请教下，你的预训练数据、微调数据、还有训练迭代到什么程度才停呀？
